netket/sampler/rules/hamiltonian.py

```diff
@@ -114,3 +116,70 @@ def _choose(vp, sections, rand_vec, out, w):
         out[i] = vp[n_rand]
         w[i] = math.log(s - low_range)
         low_range = s
+
+
+@struct.dataclass
+class HamiltonianRuleJax(HamiltonianRule):
+    """
+    Rule proposing moves according to the terms in an operator.
+
+    In this case, the transition matrix is taken to be:
+
+    .. math::
+
+       T( \\mathbf{s} \\rightarrow \\mathbf{s}^\\prime) = \\frac{1}{\\mathcal{N}(\\mathbf{s})}\\theta(|H_{\\mathbf{s},\\mathbf{s}^\\prime}|),
+
+    This rule only works with operators which are written in jax.
+    """
+
+    def transition(self, _0, _1, _2, _3, key, x):
+
+        xp, mels = self.operator.get_conn_padded(x)
+
+        n_conn = self.operator.n_conn(x)
+
+        # def _n_conn(mels):
+        #     nonzeros = jnp.abs(mels) > 0
+        #     return nonzeros.sum(axis=-1)
+        # n_conn = _n_conn(mels)
+
+        rand_i = jax.random.randint(key, shape=(x.shape[0],), minval=0, maxval=n_conn)
+
+        # we need to shift rand_i so that we only select the nonzeros
+        #
+        # if we were to assume that the nonzero mels are gathered at the beginning of each
+        # row we could avoid this, i.e. just do
+        # x_proposed = xp[jnp.arange(xp.shape[0]), rand_i]
+        #
+        # instead of actually shifting we build a mask for the selected element in each row
+        # and sum over the zeros at the end
+
+        # TODO let the operator supply the nonzeros or nonzero_i_plus1 directly
+        nonzeros = jnp.abs(mels) > 0
```

Comment:
alternatively it would be even more useful if the operator already gave us a mask telling us where the nonzeros are, not just how many there are, then we could avoid checking here

Reply:
can you propose an API about this that would make sense? Would this be useful somewhere else as well?

Reply:
the simplest thing I could think of is to add two flags to get_conn_padded, return_n_conn and return_mask, then the operator can decide what he wants to return, with None being not available.
Then depending of the application we request what we want work with what we get.

could be useful in the future if we have operators which have very non-uniform number of connected elements, or really dont know a priory how many there are, e.g. in chemistry. We could then do the computations in a loop and stop when we reached the number of connected elements actually found.
 Also would be useful for checking we did not find more connected elements than _max_conn_size (applies to e.g. chemistry)

I think we can save it for a later PR though, once we have those operators in jax it will be more clear.